### PR TITLE
Polish CWE-322: detect and exclude cases where host-checking is optional

### PR DESCRIPTION
**Pull request by smowton**

### CODE REVIEW

File: ql/src/experimental/CWE-322/InsecureHostKeyCallback.ql [hunk 1]
```diff
@@ -31,23 +41,57 @@ class InsecureHostKeyCallbackFunc extends DataFlow::Node {
   }
 }
 
+/**
+ * Identifies HostKeyCallbackFunc instances that reach CoientConfig.HostKeyCallback fields.
```
Comment (intrigus-lgtm) [inline ql/src/experimental/CWE-322/InsecureHostKeyCallback.ql:45]:
```suggestion
 * Identifies `HostKeyCallbackFunc` instances that reach `ClientConfig.HostKeyCallback` fields.
```

File: ql/src/experimental/CWE-322/InsecureHostKeyCallback.ql [hunk 2]
```diff
@@ -10,14 +10,24 @@
 import go
 import DataFlow::PathGraph
 
+/** ssh.InsecureIgnoreHostKey, which allows connecting to any host regardless of its host key. */
```
Comment (intrigus-lgtm) [inline ql/src/experimental/CWE-322/InsecureHostKeyCallback.ql:13]:
```suggestion
/** The `ssh.InsecureIgnoreHostKey` function, which allows connecting to any host regardless of its host key. */
```

File: ql/src/experimental/CWE-322/InsecureHostKeyCallback.ql (hunk 1 above)
Comment (owen-mc) [reply ql/src/experimental/CWE-322/InsecureHostKeyCallback.ql:45]: Ha, I vaguely remember thinking "Coient - what a weird word" as I read that. I didn't realise it was a typo! 😆

File: ql/src/experimental/CWE-322/InsecureHostKeyCallback.ql [hunk 3]
```diff
@@ -31,23 +41,57 @@ class InsecureHostKeyCallbackFunc extends DataFlow::Node {
   }
 }
 
+/**
+ * Identifies HostKeyCallbackFunc instances that reach `ClientConfig.HostKeyCallback` fields.
```
Comment (max-schaefer) [inline ql/src/experimental/CWE-322/InsecureHostKeyCallback.ql:45]:
```suggestion
 * A data-flow configuration for identifying `HostKeyCallbackFunc` instances that reach `ClientConfig.HostKeyCallback` fields.
```

**Comment by smowton** [comment]
@max-schaefer amended and promoted out of experimental status. Also did the same to CWE-327.

File: ql/src/Security/CWE-322/InsecureHostKeyCallback.ql [hunk 4]
```diff
@@ -0,0 +1,97 @@
+/**
+ * @name Use of insecure HostKeyCallback implementation
+ * @description Detects insecure SSL client configurations with an implementation of the `HostKeyCallback` that accepts all host keys.
+ * @kind path-problem
+ * @problem.severity error
```
Comment (max-schaefer) [inline ql/src/Security/CWE-322/InsecureHostKeyCallback.ql:5]: Needs a `@precision` now.

File: ql/src/Security/CWE-322/InsecureHostKeyCallback.ql [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+  }
+}
+
+/** A callback function value that is insecure when used as a `HostKeyCallback`, because it always returns `nil`. */
+class InsecureHostKeyCallbackFunc extends HostKeyCallbackFunc {
+  InsecureHostKeyCallbackFunc() {
+    // Either a call to InsecureIgnoreHostKey(), which we know returns an insecure callback.
+    this = any(InsecureIgnoreHostKey f).getACall()
+    or
+    // Or a callback function in the source code  (named or anonymous) that always returns nil.
```
Comment (max-schaefer) [inline ql/src/Security/CWE-322/InsecureHostKeyCallback.ql:35]:
```suggestion
    // Or a callback function in the source code (named or anonymous) that always returns nil.
```

File: ql/src/Security/CWE-322/InsecureHostKeyCallback.ql [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+
+/**
+ * A data-flow configuration for identifying `HostKeyCallbackFunc` instances that reach `ClientConfig.HostKeyCallback` fields.
+ */
+class HostKeyCallbackAssignmentConfig extends DataFlow::Configuration {
+  HostKeyCallbackAssignmentConfig() { this = "HostKeyCallbackAssignmentConfig" }
+
+  override predicate isSource(DataFlow::Node source) { source instanceof HostKeyCallbackFunc }
+
+  predicate isSink(DataFlow::Node sink, Write write) {
```
Comment (max-schaefer) [inline ql/src/Security/CWE-322/InsecureHostKeyCallback.ql:52]: Needs qldoc.

**Comment by smowton** [comment]
@max-schaefer all suggestions applied, and re: the test it already covered one case but not another, which I've added. It also turned out I'd partly broken the test without noticing :/ the results are now as expected, and I'm rerunning the LGTM tests to check if that's had any unexpected impact.

**Comment by smowton** [comment]
Btw if you look at the diff commit-by-commit you get a much nicer diff than the whole thing piled together

File: ql/test/query-tests/Security/CWE-322/vendor/golang.org/x/crypto/ssh/knownhosts/stub.go [hunk 7]
```diff
@@ -0,0 +1,12 @@
+// A simple manual stub of golang.org/x/crypto/ssh/knownhosts.New
+// See the LICENSE file for information about the licensing of the original library.
```
Comment (max-schaefer) [inline ql/test/query-tests/Security/CWE-322/vendor/golang.org/x/crypto/ssh/knownhosts/stub.go:2]: I can't see the `LICENSE` file in the diff.

Comment (smowton) [reply ql/test/query-tests/Security/CWE-322/vendor/golang.org/x/crypto/ssh/knownhosts/stub.go:2]: It already existed at the top of the golang.org tree. This PR moves it: `ql/test/{experimental => query-tests/Security}/CWE-322/vendor/golang.org/LICENSE`

Comment (max-schaefer) [reply ql/test/query-tests/Security/CWE-322/vendor/golang.org/x/crypto/ssh/knownhosts/stub.go:2]: :+1:

**Pull request merged**